pkg/common/vsphere/cns_manager.go

```diff
@@ -0,0 +1,149 @@
+package vsphere
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/vmware/govmomi/cns"
+	cnstypes "github.com/vmware/govmomi/cns/types"
+	"github.com/vmware/govmomi/object"
+	"sync"
+)
+
+type CnsManager interface {
+	CreateVolume(ctx context.Context, createSpecList []cnstypes.CnsVolumeCreateSpec) (*object.Task, error)
+	DeleteVolume(ctx context.Context, volumeIDList []cnstypes.CnsVolumeId, deleteDisk bool) (*object.Task, error)
+	QueryVolume(ctx context.Context, queryFilter cnstypes.CnsQueryFilter) (*cnstypes.CnsQueryResult, error)
+	ResetManager(ctx context.Context, vcenter *VirtualCenter, cnsClient *cns.Client)
+}
+
+var (
+	cnsManagerInstance     *defaultCnsManager
+	cnsManagerInstanceLock sync.Mutex
+)
+
+// defaultCnsManager provides functionality to manage volumes.
+type defaultCnsManager struct {
+	logger        logrus.FieldLogger
+	virtualCenter *VirtualCenter
+	cnsClient     *cns.Client
+}
+
+// GetCnsManager returns the Manager instance.
+func GetCnsManager(ctx context.Context, vc *VirtualCenter, cnsClient *cns.Client, logger logrus.FieldLogger) (CnsManager, error) {
+	cnsManagerInstanceLock.Lock()
+	defer cnsManagerInstanceLock.Unlock()
+	if cnsManagerInstance != nil {
+		logger.Infof("Retrieving existing defaultCnsManager...")
+		cnsManagerInstance.virtualCenter = vc
+		cnsManagerInstance.cnsClient = cnsClient
+		logger.Infof("Updated existing defaultCnsManager...")
+		return cnsManagerInstance, nil
+	}
+	logger.Infof("Initializing new defaultCnsManager...")
+	cnsManagerInstance = &defaultCnsManager{
+		virtualCenter: vc,
+		cnsClient:     cnsClient,
+		logger:        logger,
+	}
+	return cnsManagerInstance, nil
+}
+
+func (m *defaultCnsManager) CreateVolume(ctx context.Context, createSpecList []cnstypes.CnsVolumeCreateSpec) (*object.Task, error) {
+	log := m.logger
+	invokeCount := 0
+	var createVolumeTask *object.Task
+	cachedCnsClient := m.cnsClient
+	var err error
+	for invokeCount <= DefaultAuthErrorRetryCount {
+		invokeCount++
+		createVolumeTask, err = cachedCnsClient.CreateVolume(ctx, createSpecList)
+		if err != nil {
+			if invokeCount > DefaultAuthErrorRetryCount {
+				log.Errorf("Exceeded retry count for auth errors, err: %v", err)
+				return nil, err
+			}
+			if CheckForVcAuthFaults(err, log) {
+				// On auth fault re-initiate a connect and try again.
+				_, cachedCnsClient, err = m.virtualCenter.Connect(ctx)
```

Comment:
I don't see m.cnsClient getting updated on this path.

Reply:
This is to handle temporary issues, for example, session was killed from VC, in that it just uses the older set of credentials to re create clients and attempt to invoke the api again. The password rotation changes will explicitly come in from `ResetManager`, only `ResetManager` is able to update the vsom and vcenter associated with the VslmManager.